### PR TITLE
[Merged by Bors] - Emit errors as K8s events

### LINKED ISSUE #376: Write Kubernetes Events for errors during reconciliation
We would like to emit Kubernetes events for all errors, please see the epic for details https://github.com/stackabletech/issues/issues/158

### PR DESCRIPTION
**Pull request by teozkr**
## Description

Fixes #376, depends on https://github.com/stackabletech/operator-rs/pull/322

## Review Checklist
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator-binary/src/main.rs [hunk 1]
```diff
@@ -140,8 +141,33 @@ async fn main() -> anyhow::Result<()> {
...
                     },
                     znode_controller::error_policy,
-                    Context::new(znode_controller::Ctx { client }),
-                );
+                    Context::new(znode_controller::Ctx {
+                        client: client.clone(),
+                    }),
+                )
+                .or_else(|err| async {
+                    if let controller::Error::ReconcilerFailed(err, znode) = &err {
```
Comment (teozkr) [inline rust/operator-binary/src/main.rs:149]: Should generalize this logic (and move it into operator-rs?)

File: rust/operator-binary/src/main.rs [hunk 2]
```diff
@@ -140,8 +141,33 @@ async fn main() -> anyhow::Result<()> {
...
+                            Reporter {
+                                controller: "zookeeperznode.zookeeper.stackable.tech".to_string(),
+                                instance: None,
+                            },
+                            znode.clone().into(),
+                        );
+                        let _ = recorder
+                            .publish(Event {
+                                type_: EventType::Warning,
+                                reason: "ReconcilerFailed".to_string(),
```
Comment (teozkr) [inline rust/operator-binary/src/main.rs:161]: Can we get a better reason from the error variant?

File: rust/operator-binary/src/main.rs [hunk 3]
```diff
@@ -140,8 +141,33 @@ async fn main() -> anyhow::Result<()> {
...
+                        client: client.clone(),
+                    }),
+                )
+                .or_else(|err| async {
+                    if let controller::Error::ReconcilerFailed(err, znode) = &err {
+                        let recorder = Recorder::new(
+                            client.as_kube_client(),
+                            Reporter {
+                                controller: "zookeeperznode.zookeeper.stackable.tech".to_string(),
+                                instance: None,
```
Comment (teozkr) [inline rust/operator-binary/src/main.rs:154]: Track instance somehow

File: rust/operator-binary/src/main.rs (hunk 1 above)
Comment (teozkr) [reply rust/operator-binary/src/main.rs:149]: Also need to do this for `ZookeeperCluster`

File: Cargo.toml [hunk 4]
```diff
@@ -1,2 +1,11 @@
...
 members = ["rust/crd", "rust/operator-binary"]
+
+[patch.crates-io]
+# Depends on https://github.com/kube-rs/kube-rs/pull/815
+kube = { git = "https://github.com/kube-rs/kube-rs.git" }
+kube-core = { git = "https://github.com/kube-rs/kube-rs.git" }
+kube-runtime = { git = "https://github.com/kube-rs/kube-rs.git" }
+# kube = { path = "../../kube-rs/kube" }
+# kube-core = { path = "../../kube-rs/kube-core" }
+# kube-runtime = { path = "../../kube-rs/kube-runtime" }
```
Comment (teozkr) [inline Cargo.toml:11]: Blocked on kube-rs 0.69.0

**Comment by teozkr** [comment]
There's still some work to be done docs-wise, but I think it's in a state where I'd like feedback on both the structure and the reported errors now.

File: Cargo.toml [hunk 5]
```diff
@@ -1,2 +1,6 @@
 [workspace]
 members = ["rust/crd", "rust/operator-binary"]
+
+[patch."git+https://github.com/stackabletech/operator-rs.git"]
+# Depends on https://github.com/stackabletech/operator-rs/pull/322
+stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "feature/error-events" }
```
Comment (maltesander) [inline Cargo.toml:6]: Remove and go for operator-rs 0.11.0?

**Review by maltesander — COMMENTED** [review]
Tested and works. I like it. The errors may look strange somewhat in the events if it is a long chain but i think it is fine for now?

File: Cargo.toml (hunk 5 above)
Comment (teozkr) [reply Cargo.toml:6]: Yes.

**Review by maltesander — APPROVED** [review]
LGTM.

**Pull request closed without merging**